sources/ElkArte/Graphics/Gd2.php

```diff
@@ -0,0 +1,494 @@
+		}
+
+		$this->_setImage($image);
+
+		return true;
+	}
+
+	/**
+	 * Resize an image proportionally to fit within the defined max_width and max_height limits
+	 *
+	 * What it does:
+	 *
+	 * - Will do nothing to the image if the file fits within the size limits
+	 *
+	 * @param int|null $max_width The maximum allowed width
+	 * @param int|null $max_height The maximum allowed height
+	 * @param bool $strip if to remove the images Exif data (GD will always)
+	 * @param bool $force_resize = false Whether to override defaults and resize it
+	 *
+	 * @return bool Whether resize was successful.
+	 */
+	public function resizeImage($max_width = null, $max_height = null, $strip = true, $force_resize = false)
+	{
+		$success = false;
+
+		// No image, no further
+		if (empty($this->_image))
+		{
+			return $success;
+		}
+
+		$src_width = $this->_width;
+		$src_height = $this->_height;
+
+		// Allow for a re-encode to the same size
+		$max_width = $max_width === null ? $src_width : $max_width;
+		$max_height = $max_height === null ? $src_height : $max_height;
+
+		// Determine whether to resize to max width or to max height (depending on the limits.)
+		if (!empty($max_width) && !empty($max_height))
+		{
+			$image_ratio = $this->_width / $this->_height;
+			$requested_ratio = $max_width / $max_height;
+
+			if ($requested_ratio > $image_ratio)
+			{
+				$dst_width = max(1, $max_height * $image_ratio);
+				$dst_height = $max_height;
+			}
+			else
+			{
+				$dst_width = $max_width;
+				$dst_height = max(1, $max_width / $image_ratio);
+			}
+
+			// Don't bother resizing if it's already smaller...
+			if (!empty($dst_width) && !empty($dst_height) && ($dst_width < $src_width || $dst_height < $src_height || $force_resize))
+			{
+				$dst_img = imagecreatetruecolor($dst_width, $dst_height);
+				$this->_createCanvas($dst_img);
+
+				// Resize it!
+				$success = imagecopyresampled($dst_img, $this->_image, 0, 0, 0, 0, $dst_width, $dst_height, $src_width, $src_height);
+			}
+			else
+			{
+				$dst_img = $this->_image;
+				$success = true;
+			}
+		}
+		else
+		{
+			$dst_img = $this->_image;
+		}
+
+		// Update all settings to the converted image
+		$this->_setImage($dst_img);
+
+		return $success;
+	}
+
+	/**
+	 * Create a transparent true image canvas to place our image on
+	 *
+	 * @param resource $dst_img
+	 */
+	protected function _createCanvas($dst_img)
+	{
+		// Make a true color image, because it just looks better for resizing.
+		imagesavealpha($dst_img, true);
+		$color = imagecolorallocatealpha($dst_img, 255, 255, 255, 127);
+		imagefill($dst_img, 0, 0, $color);
+	}
+
+	/**
+	 * Output the image resource to a file in a chosen format
+	 *
+	 * @param string $file_name where to save the image, if null output to screen
+	 * @param int $preferred_format jpg,png,gif, etc
+	 * @param int $quality the jpg image quality
+	 *
+	 * @return bool
+	 */
+	public function output($file_name = null, $preferred_format = IMAGETYPE_JPEG, $quality = 85)
```

Comment:
Scrutinizer says it should be string|null, though maybe the default should be an empty string because of the strict comparison afterwards?

Reply:
null outputs to the stream to the buffer, I did that vs '', my intention was for an implicit request vs just a mistake,  

I'll add the string|null BUT I bet Scrutinizer will complain about the other called methods in output that use file_name but require a string .... So I guess to clear that "error"  I'll have to drop the null and just use a blank string file_name to echo the blob.